flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/JobGraph.java

```diff
@@ -300,6 +305,32 @@ public int getNumberOfVertices() {
 		return this.taskVertices.size();
 	}
 
+	public Set<SlotSharingGroup> getSlotSharingGroups() {
+		final Set<SlotSharingGroup> slotSharingGroups = new HashSet<>();
+		for (JobVertex vertex : getVertices()) {
+			final SlotSharingGroup slotSharingGroup = vertex.getSlotSharingGroup();
+			checkNotNull(slotSharingGroup);
+
+			slotSharingGroups.add(slotSharingGroup);
+		}
+		return Collections.unmodifiableSet(slotSharingGroups);
+	}
+
+	public Set<CoLocationGroupDesc> getCoLocationGroupDescriptors() {
+		final Set<CoLocationGroup> coLocationGroups = new HashSet<>();
+		for (JobVertex vertex : getVertices()) {
+			CoLocationGroup coLocationGroup = vertex.getCoLocationGroup();
+			if (coLocationGroup != null) {
+				coLocationGroups.add(coLocationGroup);
+			}
+		}
+		final Set<CoLocationGroupDesc> coLocationGroupDescs = coLocationGroups
+			.stream()
+			.map(CoLocationGroupDesc::from)
+			.collect(Collectors.toSet());
```

Comment:
We cannot do this because `CoLocationGroupDesc` does not override `equals(...)`. And I'd like to avoid the equality comparison which may require to compare 2 lists.

Reply:
Then maybe we substitute `Set<CoLocationGroupDesc>` with `Map<AbstractID, CoLocationGroupDesc>`?
```
final Set<CoLocationGroupDesc> coLocationGroupDescs = getVertices()
    .stream()
    .map(vertex -> CoLocationGroupDesc.from(vertex.getCoLocationGroup()))
    .collect(Collectors.toMap(CoLocationGroupDesc::getId, d -> d))
    .values()
    .stream()
    .collect(Collectors.toSet());
```

It can be also in a for-loop, but it is nit anyways so I think it is ok as it is.
Maybe, at least a comment why we do it like this: `because equals ...`.

Reply:
A comment is added to explain that.
The loops are also reworked to be in stream format, but in a bit different way.